Sizeof should fail on incomplete types

xbauch opened the pull request:
And so we check for it when type-checking the `sizeof` expression.

Note: previously, the `size-of-expr` function would return zero (sum of the struct components sizes).

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented:
May fix #5108.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 884c36074a850d0383497a53a0e0ed7889cfdb61).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143254568

chrisr-diffblue commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -951,6 +951,16 @@ void c_typecheck_baset::typecheck_expr_sizeof(exprt &expr)
+      to_struct_type(follow_tag(to_struct_tag_type(type))).is_incomplete())
+    {
+      error().source_location = expr.source_location();
+      error() << "calling sizeof for incomplete type: " << to_string(type)
```
Looks good in principal, I just have one suggestion, and one query:
* I think you could pretty much copy the compiler error message from gcc for the error message here:
```
gcc cbmc-test.c
cbmc-test.c:8:31: error: invalid application of 'sizeof' to an incomplete type
      'struct foo'
  struct foo* thefoo = malloc(sizeof(struct foo));
```
* This works for incomplete *structs* - but it won't catch other incomplete types? (e.g. unions, arrays with no dimensions, etc?)

xbauch replied:
I've change the message and extended the check to unions and enums. Arrays are weird: at the time of typechecking we already see variables of type `int[]` as if having type `int`.

chrisr-diffblue replied:
Hmm, so does that mean you don't even know they are an array at that point? Also, what's the difference at that point between a variable of type `int[]` and a variable of type `int[10]` - does one have a length attribute somewhere? If you can know it's an array, and know it's length then it's a complete type. If you don't know it's length, then it's incomplete.

xbauch replied:
My bad: I was looking at a different `sizeof`. Now it should work for enums and arrays as well.

chrisr-diffblue reviewed (approved):
Looks good, many thanks :-)

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -951,6 +951,21 @@ void c_typecheck_baset::typecheck_expr_sizeof(exprt &expr)
   {
+    if(
+      (type.id() == ID_struct_tag &&
+       to_struct_type(follow_tag(to_struct_tag_type(type))).is_incomplete()) ||
```
No need for the `to_struct_type` etc.

xbauch replied:
Done.

kroening commented:
I'd stick the tests into one directory, say "incomplete-sizeof".

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a41a250b21c34a32b3e63038bfa8e49531cd275a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143432884

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 5564772dcb6ec4fc3171c1650166305a5375c94f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143445520
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 81499c5d30d3d759e19e6e19bd7b6cb25074e182).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143449884

petr-bauch committed:
Sizeof should fail on incomplete types

And so we check for it when type-checking the `sizeof` expression. Extends to
incomplete enums, structs, unions, and arrays (zero-length).

Note: previously, the `size-of-expr` function would return zero (sum of the
struct components sizes).

hannes-steffenhagen-diffblue reviewed (approved):
LGTM

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 59c6d8fb974384f3df14088122ecd5df2e6349ee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/143553136
